system/Database/BaseResult.php

```diff
@@ -130,7 +130,7 @@ public function getCustomResultObject(string $className)
 			return $this->customResultObject[$className];
 		}
 
-		if (is_bool($this->resultID) || ! $this->resultID || $this->numRows === 0)
+		if (is_bool($this->resultID) || ! $this->resultID)
```

Comment:
Are you sure? I would point out that numRows will be NULL until you call getNumRows for the first time so this check will *never* be true unless you've done so. For SQLite3, I *think* there is a potential for unwanted looping.  For example, this controller method:
```php
	function test4() {
		$db = \Config\Database::connect("tests");
		echo get_class($db) . "\n";
		
		$query = $db->query("SELECT * FROM db_job");
		$num = $query->getNumRows();
		$rows = $query->getResultObject();
	}

```
That call to `getNumRows()` falls back on `BaseResult::getNumRows()` which ends up calling `getResultArray()` to fetch all the results (see [line 555 in BaseResult](https://github.com/sneakyimp/CodeIgniter4/blob/num-rows-signed/system/Database/BaseResult.php#L555) ) and count them. Then, when you call `getResultObject()` the code will have to loop through all the `resultArray` to cast each row of resultArray as an object (see [line 248 in BaseResult](https://github.com/sneakyimp/CodeIgniter4/blob/num-rows-signed/system/Database/BaseResult.php#L248) ).

I would add that if you *never* call getNumRows -- which *must be the case currently for all existing CI4 projects* -- that the check `$this->numRows === 0` will *never* evaluate to `true`. Any time resultID is some iterable set this is a useless check *unless you happen to call getNumRows first.* I don't think this check accomplishes much, because it only helps IF:
* You have previously called getNumRows
* the query actually returned an result set that happens to have zero records.

Conversely, if your resultID is not boolean and not empty and you have not called getNumRows (probably pretty common) then you are adding a useless check that chews up about 55-60 nanoseconds (a few hundred clock cycles?) *every time this function runs.* 

Lastly, assuming the conditions are just right for this check to be helpful -- but we fail to perform the check -- that would mean we are dealing with an empty record set and have called getNumRows previously. In this case we would only have a few if clauses to check:
1) if resultArray/resultObject (false)
2) `is_null($this->rowData) || $this->dataSeek();`
3) `while ($row = $this->fetchAssoc())`

I did a bit of simple performance profiling on those three checks and they run in an average of 7.4 microseconds. Do developers fetch an iterable-but-empty record set *after* calling getNumRows often enough that saving these 7.4 microseconds is worthwhile? It seems much more common to me that developers will run a query and fetch a result set without calling getNumRows in which case we are paying that penalty of 55-60 nano seconds.

Let me know your thoughts.

Reply:
Yes, unwanted looping is something we can expect for SQLite3 driver for sure.

Since you have done some profiling I'm good with not adding it back. Thank you 👍